projects/plugins/jetpack/modules/masterbar/admin-menu/class-base-admin-menu.php

```diff
@@ -455,6 +455,74 @@ public function override_svg_icons() {
 		}
 	}
 
+	/**
+	 * Hide menus that are unauthorized and don't have visible submenus.
+	 *
+	 * This must be done at the end of menu and submenu manipulation in order to avoid performing this check each time
+	 * the submenus are altered.
+	 */
+	public function hide_unauthorized_menus() {
+		global $menu, $submenu;
+
+		foreach ( $menu as $menu_index => $menu_item ) {
+			$has_submenus = isset( $submenu[ $menu_item[2] ] );
+
+			// Skip if the menu doesn't have submenus.
+			if ( ! $has_submenus ) {
+				continue;
+			}
+
+			// If the first submenu item is hidden then we should also hide the parent.
+			// Since the submenus are ordered by self::HIDE_CSS_CLASS (hidden submenus should be at the end of the array),
+			// we can say that if the first submenu is hidden then we should also hide the menu.
+			$first_submenu_item       = array_values( $submenu[ $menu_item[2] ] )[0];
+			$is_first_submenu_visible = $this->is_item_visible( $first_submenu_item );
+
+			// if the user does not have access to the menu and the first submenu is hidden, then hide the menu.
+			if ( ! current_user_can( $menu_item[1] ) && ! $is_first_submenu_visible ) {
+				// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+				$menu[ $menu_index ][4] = self::HIDE_CSS_CLASS;
+			}
+
+			// if the menu has the same slug as the first submenu then hide the submenu.
+			if ( $menu_item[2] === $first_submenu_item[2] && ! $is_first_submenu_visible ) {
+				// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+				$menu[ $menu_index ][4] = self::HIDE_CSS_CLASS;
+			}
+		}
+	}
+
+	/**
+	 * Sort the hidden submenus by moving them at the end of the array in order to avoid WP using them as default URLs.
+	 *
+	 * This operation has to be done at the end of submenu manipulation in order to guarantee that the hidden submenus
+	 * are at the end of the array.
+	 */
+	public function sort_hidden_submenus() {
+		global $submenu;
+
+		foreach ( $submenu as $menu_slug => $submenu_items ) {
+			foreach ( $submenu_items as $submenu_index => $submenu_item ) {
+				if ( $this->is_item_visible( $submenu_item ) ) {
+					continue;
+				}
+
+				unset( $submenu[ $menu_slug ][ $submenu_index ] );
+				// phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+				$submenu[ $menu_slug ][] = $submenu_item;
+			}
+		}
+	}
+
+	/**
+	 * Check if the given item is visible or not in the admin menu.
+	 *
+	 * @param array $item A menu or submenu array.
+	 */
+	public function is_item_visible( $item ) {
+		return ! isset( $item[4] ) || false === strpos( self::HIDE_CSS_CLASS, $item[4] );
```

Comment:
Non-blocking: We could refactor the `has_visible_items` function to use this new function.

Reply:
Sure, great idea! 👍